### PR TITLE
`Integrated code lifecycle`:  Add repository buttons for exams

### PR DESCRIPTION
**Pull request by mateusmm01**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Currently, users can't see the repositories in the Exam Summary View. This should be enabled not only for students but also instructors, etc.

### Description

Mainly, new routes have been added, as well as an open repository button in `programming-exam-summary.component`. Furthermore the logic in the server side has to be changed to allow students to open their repositories with READ access also after the exam due date.

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Check in the Exam Summary that you can open the repository of your programming exercise and everything works as expected
4. As an instructor check that you can see the repository in the exam summary for different students.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance

#### Exam Mode Test
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exam-summary.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6202/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/summary/exercises/programming-exam-summary/programming-exam-summary.component.ts.html) | 100% | ✅ |
| [exercise-details-student-actions.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6202/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/exercise-details/exercise-details-student-actions.component.ts.html) | 88.09% | ✅ ❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [RepositoryAccessService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6202/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/RepositoryAccessService.html) | 97% | ✅ ❌ |
| [ResultService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6202/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ResultService.html) | 92% | ✅ ❌ |
| [PlagiarismService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6202/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.plagiarism/PlagiarismService.html) | 91% | ✅ ❌ |
| [ProgrammingExerciseParticipationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6202/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ProgrammingExerciseParticipationResource.html) | 82% | ✅ ❌ |

### Screenshots

Screenshot 2024-03-10 at 09 54 46

## Summary by CodeRabbit

- **New Features**
	- Introduced a new variable `exerciseStartDate` and refined access control based on it.
- **Bug Fixes**
	- Enhanced security by adjusting access control logic related to exercise start dates.
- **Tests**
	- Updated integration tests to reflect changes in exercise timing, file access, plagiarism cases, and exam constraints.
	- Added tests for creating repository URLs in `ExerciseDetailsStudentActionsComponent`.

### CODE REVIEW

File: src/main/webapp/app/exam/manage/exam-management.route.ts [hunk 1]
```diff
@@ -299,6 +303,42 @@ export const examManagementRoute: Routes = [
         },
         canActivate: [UserRouteAccessService],
     },
+    {
+        path: ':examId/student-exams/:studentExamId/summary/exercises/:exerciseId/repository/:participationId',
+        component: RepositoryViewComponent,
+        data: {
+            authorities: [Authority.USER],
```
Comment (rstief) [inline src/main/webapp/app/exam/manage/exam-management.route.ts:310]: This is on the management route. Is it intended that this is accessible by users or should this be TA and higher?

File: src/main/webapp/app/exam/participate/summary/exercises/programming-exam-summary/programming-exam-summary.component.html [hunk 2]
```diff
@@ -5,6 +5,18 @@
...
+                    @if (localVCEnabled) {
+                        <a
+                            class="btn btn-primary"
+                            [class.btn-sm]="true"
+                            [routerLink]="routerLink + '/exercises/' + exercise.id + '/repository/' + participation.id"
+                            target="_blank"
+                            rel="noopener noreferrer"
+                        >
+                            <fa-icon [icon]="faCodeBranch" [fixedWidth]="true" />
+                            <span class="d-none d-md-inline" jhiTranslate="artemisApp.exerciseActions.openRepository">Open Repository</span>
```
Comment (rstief) [inline src/main/webapp/app/exam/participate/summary/exercises/programming-exam-summary/programming-exam-summary.component.html:17]:
```suggestion
                            <span class="d-none d-md-inline" jhiTranslate="artemisApp.exerciseActions.openRepository"></span>
```

**Review by rstief — COMMENTED** [review]
Code looks good overall, two small comments :)

File: src/main/webapp/app/exam/manage/exam-management.route.ts (hunk 1 above)
Comment (mateusmm01) [reply src/main/webapp/app/exam/manage/exam-management.route.ts:310]: Yes thank you :)

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 3]
```diff
@@ -54,27 +56,80 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
+        if (repositoryActionType == RepositoryActionType.WRITE && !isStudentParticipation) {
+            throw new AccessUnauthorizedException();
+        }
+
+        // Error case 4: A student tries to push into a repository that does not belong to them.
+        if (repositoryActionType == RepositoryActionType.WRITE && !((StudentParticipation) programmingParticipation).isOwnedBy(user)) {
+            throw new AccessUnauthorizedException();
+        }
+
+        // Error case 4: The student can reset the repository only before and a tutor/instructor only after the due date has passed
```
Comment (laurenzfb) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:90]:
```suggestion
        // Error case 5: The student can reset the repository only before and a tutor/instructor only after the due date has passed
```

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 4]
```diff
@@ -54,27 +56,80 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
+        if (repositoryActionType == RepositoryActionType.RESET) {
+            checkAccessRepositoryForReset(programmingParticipation, isStudent, programmingExercise);
+        }
+
+        // Error case 5: Check if the user is allowed to access the repository concerning the dates
```
Comment (laurenzfb) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:95]:
```suggestion
        // Error case 6: Check if the user is allowed to access the repository concerning the dates
```

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 5]
```diff
@@ -54,27 +56,80 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
+        if (isStudent) {
+            checkIsStudentAllowedToAccessRepositoryConcerningDates(programmingParticipation, repositoryActionType, exerciseStartDate);
+        }
+
+        // Error case 6: Check if the user is allowed to submit for the exam
```
Comment (laurenzfb) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:100]:
```suggestion
        // Error case 7: Check if the user is allowed to submit for the exam
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java [hunk 6]
```diff
@@ -303,4 +305,19 @@ public ModelAndView redirectGetParticipationRepositoryFilesForCommitsDetailsView
         return new ModelAndView("forward:/api/repository/" + participation.getId() + "/files-content/" + commitId);
     }
 
+    /**
+     * Checks if the user has access to the participation.
+     * If the exercise has not started yet and the user is a student access is denied.
```
Comment (laurenzfb) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:310]:
```suggestion
     * If the exercise has not started yet and the user is a student, access is denied.
```

**Review by laurenzfb — CHANGES_REQUESTED** [review]
Tested on TS6 and Local CI Staging. Unfortunately, there seems to be a problem when trying to create a new exam:

`2024-03-09 19:10:59.259  WARN 1 --- [nio-8080-exec-5] o.z.problem.spring.common.AdviceTraits   : Bad Request: Failed to convert value of type 'java.lang.String' to required type 'long'; nested exception is java.lang.NumberFormatException: For input string: "NaN"`

I don't know, how this is caused by your changes but the problem does not occur on `develop`. 

Concerning the code, I left some minor suggestions

File: src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCLocalCIIntegrationTest.java [hunk 7]
```diff
@@ -324,14 +324,14 @@ void testFetchPush_studentAssignmentRepository_beforeAfterStartDate() throws Exc
         localVCLocalCITestService.testFetchReturnsError(assignmentRepository.localGit, student2Login, projectKey1, assignmentRepositorySlug, FORBIDDEN);
         localVCLocalCITestService.testPushReturnsError(assignmentRepository.localGit, student2Login, projectKey1, assignmentRepositorySlug, FORBIDDEN);
 
-        // Before the start date of the exercise, students are able to fetch if (as in this case) their repository already exists. This is consistent with the behaviour of
+        // Before the start date of the exercise, students aren't able to access their repositories. This is consistent with the behaviour of
         // Bitbucket and GitLab. Usually, the exercise will be configured with a start date in the future and students will not be able to create a repository before that.
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCLocalCIIntegrationTest.java:328]:
```suggestion
        // Before the start date of the exercise, students aren't able to access their repositories. 
        // Usually, the exercise will be configured with a start date in the future and students will not be able to create a repository before that.
```

**Review by laurenzfb — APPROVED** [review]
tested on ts6. Everything works now. Students can see the repository view during and after the exam but not before. Code also lgtm

**Review by RY997 — APPROVED** [review]
Tested on TS6, both the student and instructor can view the repo in the exam summary.

**Review by marlon-luca-bu — APPROVED** [review]
Tested on TS6, works as expected 👍

**Pull request merged**